File: Solution-Java/Longest_159.java
class Longest_159 {
	// My updated solution: Longest window, refer to 3, 159, 340, 1004, 424
    public int lengthOfLongestSubstringTwoDistinct(String s) {
        int[] freq = new int[128];
        int l = 0, r = 0, distinct = 0;
        for (; r < s.length(); r++) {
            if (++freq[s.charAt(r)] == 1) distinct++;
            if (distinct > 2 && --freq[s.charAt(l++)] == 0) distinct--;
        }
        return r - l;
    }
}

File: Solution-Python/Longest_159.py
from collections import defaultdict

class Solution:
    def lengthOfLongestSubstringTwoDistinct(self, s: str) -> int:
        freq = defaultdict(int)
        l, count = 0, 0
        for c in s:
            freq[c] += 1
            if freq[c] == 1:
                count += 1
            if count > 2:
                freq[s[l]] -= 1
                if freq[s[l]] == 0:
                    count -= 1
                l += 1
        return len(s) - l